codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+
+    if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+      return failure(err)
+
+    return success()
+  else:
+    datasetMd.checkpoint.progress = index
+    return await self.deleteBatch(treeCid, datasetMd)
+
+proc superviseDatasetDeletion(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[void] {.async.} =
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    expiry = datasetMd.expiry
+    leavesCount = datasetMd.leavesCount
+
+  try:
+    if datasetMd.checkpoint.progress == 0 and datasetMd.checkpoint.timestamp == 0:
+      info "Initiating deletion of a dataset"
+    else:
+      info "Retrying deletion of a dataset", progress = datasetMd.checkpoint.progress, timestamp = datasetMd.checkpoint.timestamp
+
+    if err =? (await self.deleteBatch(treeCid, datasetMd)).errorOption:
+      error "Error occurred during deletion of a dataset", msg = err.msg
+    else:
+      info "Dataset deletion complete"
+  except CancelledError as err:
+    raise err
+  except CatchableError as err:
+    error "Unexpected error during dataset deletion", msg = err.msg, treeCid = treeCid
+
+proc listDatasetMetadata*(
+  self: DatasetMaintainer
+): Future[?!AsyncIter[(Cid, DatasetMetadata)]] {.async.} =
+  without queryKey =? createDatasetMetadataQueryKey(), err:
+    return failure(err)
+
+  without queryIter =? await query[DatasetMetadata](self.metaDs, Query.init(queryKey)), err:
+    error "Unable to execute block expirations query", err = err.msg
+    return failure(err)
+
+  without asyncQueryIter =? await queryIter.toAsyncIter(), err:
+    error "Unable to convert QueryIter to AsyncIter", err = err.msg
+    return failure(err)
+
+  let
+    filteredIter = await asyncQueryIter.filterSuccess()
+
+    datasetMdIter = await mapFilter[KeyVal[DatasetMetadata], (Cid, DatasetMetadata)](filteredIter,
+      proc (kv: KeyVal[DatasetMetadata]): Future[?(Cid, DatasetMetadata)] {.async.} =
+        without cid =? Cid.init(kv.key.value).mapFailure, err:
+          error "Failed decoding cid", err = err.msg
+          return (Cid, DatasetMetadata).none
+
+        (cid, kv.value).some
+    )
+
+  success(datasetMdIter)
+
+
+proc checkDatasets(self: DatasetMaintainer): Future[?!void] {.async.} =
+  without iter =? await self.listDatasetMetadata(), err:
+    return failure(err)
 
-  var numberReceived = 0
-  for beFut in iter:
-    let be = await beFut
-    inc numberReceived
-    await self.processBlockExpiration(be)
-    await sleepAsync(1.millis) # cooperative scheduling
+  for fut in iter:
+    let (treeCid, datasetMd) = await fut
 
-  # If we received fewer blockExpirations from the iterator than we asked for,
-  # We're at the end of the dataset and should start from 0 next time.
-  if numberReceived < self.numberOfBlocksPerInterval:
-    self.offset = 0
+    if (datasetMd.expiry < self.clock.now) and
+        (datasetMd.checkpoint.timestamp + self.retryDelay.seconds < self.clock.now):
```

Comment:
1. There can be multiple reasons for that. I.e. fail to open db file due to too many file descriptors open.
2. Because pt. 1. can happen or node can shutdown mid point dataset deletion.